src/neptune/new/internal/streams/std_stream_capture_logger.py

```diff
@@ -28,41 +29,51 @@ def __init__(self, container: MetadataContainer, attribute_name: str, stream: Te
         self.stream = stream
         self._thread_local = threading.local()
         self.enabled = True
+        self._log_data_queue = Queue()
+        self._finished_logging = threading.Event()
+        self._logging_thread = threading.Thread(target=self.__proces_logs, daemon=True)
+        self._logging_thread.start()
 
     def write(self, data: str):
-        if not hasattr(self._thread_local, "inside_write"):
-            self._thread_local.inside_write = False
-
         self.stream.write(data)
-        if self.enabled and not self._thread_local.inside_write:
-            try:
-                self._thread_local.inside_write = True
-                self._logger.log(data)
-            finally:
-                self._thread_local.inside_write = False
+        self._log_data_queue.put_nowait(data)
 
     def __getattr__(self, attr):
         return getattr(self.stream, attr)
 
-    def close(self):
+    def close(self, wait_for_all_logs=True):
+        if not wait_for_all_logs:
+            self.enabled = False
+        self._log_data_queue.put_nowait(None)
+        self._finished_logging.wait()
```

Comment:
We should probably `.join()` background thread anyway, so we can wait for join instead of an event.